raft/src/test/java/org/apache/kafka/raft/EndpointsTest.java

```diff
@@ -0,0 +1,489 @@
+                address1
+        );
+
+        assertEquals(expectedString, endpoints.toString());
+    }
+
+    @Test
+    void testToBeginQuorumEpochRequestWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        BeginQuorumEpochRequestData.LeaderEndpointCollection leaderEndpoints = endpoints.toBeginQuorumEpochRequest();
+
+        assertEquals(1, leaderEndpoints.size());
+
+        Iterator<BeginQuorumEpochRequestData.LeaderEndpoint> iterator = leaderEndpoints.iterator();
+        BeginQuorumEpochRequestData.LeaderEndpoint endpoint = iterator.next();
+        assertEquals(ListenerName.normalised("listener").value(), endpoint.name());
+        assertEquals("localhost", endpoint.host());
+        assertEquals(9092, endpoint.port());
+    }
+
+    @Test
+    void testToBeginQuorumEpochRequestWithEmptyEndpoint() {
+        Endpoints endpoints = Endpoints.empty();
+
+        BeginQuorumEpochRequestData.LeaderEndpointCollection leaderEndpoints = endpoints.toBeginQuorumEpochRequest();
+
+        assertEquals(0, leaderEndpoints.size());
+    }
+
+    @Test
+    void testEmpty() {
+        Endpoints emptyEndpoints = Endpoints.empty();
+
+        assertEquals(Endpoints.fromInetSocketAddresses(Collections.emptyMap()), emptyEndpoints);
+    }
+
+    @Test
+    void testFromInetSocketAddressesWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        assertEquals(1, endpoints.size());
+    }
+
+    @Test
+    void testFromVotersRecordEndpointsWithEndpoint() {
+        Map<ListenerName, InetSocketAddress> endpointMap = new HashMap<>();
+        endpointMap.put(
+                ListenerName.normalised("listener"),
+                InetSocketAddress.createUnresolved("localhost", 9092)
+        );
+        Endpoints endpoints = Endpoints.fromInetSocketAddresses(endpointMap);
+
+        List<VotersRecord.Endpoint> votersEndpoints = new ArrayList<>();
+        votersEndpoints.add(
+                new VotersRecord.Endpoint()
+                        .setName("listener")
+                        .setHost("localhost")
+                        .setPort(9092)
+        );
+
+        Endpoints createdEndpoints = Endpoints.fromVotersRecordEndpoints(votersEndpoints);
+
+        assertEquals(endpoints, createdEndpoints);
+    }
+
+    @Test
+    void testFromVotersRecordEndpointsWithEmptyEndpoint() {
+        List<VotersRecord.Endpoint> votersEndpoints = new ArrayList<>();
```

Comment:
Can't `votersEndpoints` be `Collections.emptyList` instead? Then  `createdEndpoints` can be `Endpoints.fromVotersRecordEndpoints(Collections.emptyList());`

Reply:
updated